Sources/FoundationXML/XMLParser.swift

```diff
@@ -532,57 +539,54 @@ open class XMLParser : NSObject {
         return result
     }
 
-    internal func parseFromStream() -> Bool {
+    internal func parseFrom(_ stream : InputStream) -> Bool {
         var result = true
-        XMLParser.setCurrentParser(self)
-        defer { XMLParser.setCurrentParser(nil) }
-        if let stream = _stream {
-            stream.open()
-            defer { stream.close() }
-            let buffer = malloc(_chunkSize)!.bindMemory(to: UInt8.self, capacity: _chunkSize)
-            defer { free(buffer) }
-            var len = stream.read(buffer, maxLength: _chunkSize)
-            if len != -1 {
-                while len > 0 {
-                    let data = Data(bytesNoCopy: buffer, count: len, deallocator: .none)
-                    result = parseData(data)
-                    len = stream.read(buffer, maxLength: _chunkSize)
-                }
-            } else {
+
+        guard let buffer = malloc(_chunkSize)?.bindMemory(to: UInt8.self, capacity: _chunkSize) else { return false }
+        defer { free(buffer) }
+
+        stream.open()
+        defer { stream.close() }
+        parseLoop: while result {
+            switch stream.read(buffer, maxLength: _chunkSize) {
+            case let len where len > 0:
+                let data = Data(bytesNoCopy: buffer, count: len, deallocator: .none)
+                result = parseData(data)
+            case 0:
+                result = parseData(Data(), lastChunkOfData: true)
+                break parseLoop
+            default: // See SR-13516, should be `case ..<0:`
                 result = false
-            }
-        } else if var data = _data {
-            let buffer = malloc(_chunkSize)!.bindMemory(to: UInt8.self, capacity: _chunkSize)
-            defer { free(buffer) }
-            var range = NSRange(location: 0, length: min(_chunkSize, data.count))
-            while result {
-                let chunk = data.withUnsafeMutableBytes { (rawBuffer: UnsafeMutableRawBufferPointer) -> Data in
-                    let ptr = rawBuffer.baseAddress!.advanced(by: range.location)
-                    return Data(bytesNoCopy: ptr, count: range.length, deallocator: .none)
-                }
-                result = parseData(chunk)
-                if range.location + range.length >= data.count {
-                    break
+                if _parserError == nil {
+                    _parserError = stream.streamError
                 }
-                range = NSRange(location: range.location + range.length, length: min(_chunkSize, data.count - (range.location + range.length)))
+
+                break parseLoop
             }
-        } else {
-            result = false
         }
+
         return result
     }
     
     // called to start the event-driven parse. Returns YES in the event of a successful parse, and NO in case of error.
     open func parse() -> Bool {
-        return parseFromStream()
+        XMLParser.setCurrentParser(self)
+        defer { XMLParser.setCurrentParser(nil) }
```

Comment:
If I start another parser from a parser callback, will this be reentrant correctly?

Reply:
As far as I understand the it: No it will not.

On the other side: That’s no difference from the current code. (old: parse → parseFromStream → setCurrentParser → parseData → callback, new: parse → setCurrentParser [→ parseFromStream] → parseData → callback)